File: microfi-mobile/lib/core/collection_chain_codec.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

/// Offline Field Collection Security Algorithm v1.1 §5/§12 hash-chain codec — the Dart mirror of
/// microfi-core's CollectionChainCodec.java. Must produce byte-identical output to it; see
/// test/core/collection_chain_codec_test.dart, which shares fixed vectors with
/// CollectionChainCodecTest.java on the server to catch any drift between the two.
///
/// An explicit, fixed-order delimited string rather than JSON — same reasoning as
/// QrReceiptSigner's _canonicalString. lat/lon are fixed to 6 decimal places and the timestamp is
/// the raw UTC epoch millisecond integer (not an ISO-8601 string) specifically to remove any
/// dependency on Dart's and Java's default number/date formatting agreeing byte-for-byte.
class CollectionChainCodec {
  /// The previousHash value for the first record in a chain (counter == 1) — mirrors the spec's GENESIS_HASH.
  static const genesisHash = 'GENESIS';

  static String canonicalString({
    required String installationId,
    required String deviceTxId,
    required String clientId,
    required int amountXaf,
    required double lat,
    required double lon,
    required int collectedAtEpochMilli,
    required int counter,
    required String previousHash,
  }) {
    return [
      installationId,
      deviceTxId,
      clientId,
      amountXaf.toString(),
      lat.toStringAsFixed(6),
      lon.toStringAsFixed(6),
      collectedAtEpochMilli.toString(),
      counter.toString(),
      previousHash,
    ].join('|');
  }

  /// Base64(SHA-256(canonicalString)).
  static String computeHash({
    required String installationId,
    required String deviceTxId,
    required String clientId,
    required int amountXaf,
    required double lat,
    required double lon,
    required int collectedAtEpochMilli,
    required int counter,
    required String previousHash,
  }) {
    final canonical = canonicalString(
      installationId: installationId,
      deviceTxId: deviceTxId,
      clientId: clientId,
      amountXaf: amountXaf,
      lat: lat,
      lon: lon,
      collectedAtEpochMilli: collectedAtEpochMilli,
      counter: counter,
      previousHash: previousHash,
    );
    return base64Encode(sha256.convert(utf8.encode(canonical)).bytes);
  }

  /// Base64(HMAC-SHA256(installationSecret, currentHash)).
  static String computeSignature({required String installationSecretBase64, required String currentHash}) {
    final secretBytes = base64Decode(installationSecretBase64);
    final signatureBytes = Hmac(sha256, secretBytes).convert(utf8.encode(currentHash)).bytes;
    return base64Encode(signatureBytes);
  }
}

File: microfi-core/src/main/java/com/microfi/transactions/service/CollectionChainCodec.java
package com.microfi.transactions.service;

import org.springframework.stereotype.Component;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;
import java.util.Locale;
import java.util.UUID;

/**
 * Offline Field Collection Security Algorithm v1.1 §5/§12 — the hash-chain codec. Must produce
 * byte-identical output to its Dart counterpart (microfi-mobile/lib/core/collection_chain_codec.dart);
 * see that file's tests, which share fixed vectors with {@code CollectionChainCodecTest} here to
 * catch any drift between the two implementations.
 * <p>
 * An explicit, fixed-order delimited string rather than re-serializing to JSON for hashing/signing
 * — same reasoning as {@code QrReceiptSigner._canonicalString} on the mobile side: avoids any
 * dependency on map/JSON key-ordering staying stable. Two deliberate departures from a naive
 * "just stringify everything the same way" approach, both to remove cross-language ambiguity
 * rather than trust Dart's and Java's default number/date formatting to agree byte-for-byte:
 * lat/lon are fixed to 6 decimal places (~11cm precision, ample for this purpose), and the
 * timestamp is the raw UTC epoch millisecond integer, not an ISO-8601 string (whose fractional-
 * second formatting differs between {@code Instant#toString()} and Dart's {@code
 * toIso8601String()}).
 */
@Component
public class CollectionChainCodec {

    /** The previousHash value for the first record in a chain (counter == 1) — mirrors the spec's GENESIS_HASH. */
    public static final String GENESIS_HASH = "GENESIS";

    public record ChainInput(String installationId, String deviceTxId, UUID clientId, long amountXaf,
                              double lat, double lon, long collectedAtEpochMilli, long counter, String previousHash) {
    }

    public String canonicalString(ChainInput input) {
        return String.join("|",
                input.installationId(),
                input.deviceTxId(),
                input.clientId().toString(),
                String.valueOf(input.amountXaf()),
                String.format(Locale.ROOT, "%.6f", input.lat()),
                String.format(Locale.ROOT, "%.6f", input.lon()),
                String.valueOf(input.collectedAtEpochMilli()),
                String.valueOf(input.counter()),
                input.previousHash());
    }

    /** {@code Base64(SHA-256(canonicalString))}. */
    public String computeHash(ChainInput input) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hashBytes = digest.digest(canonicalString(input).getBytes(StandardCharsets.UTF_8));
            return Base64.getEncoder().encodeToString(hashBytes);
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 not available", e);
        }
    }

    /** {@code Base64(HMAC-SHA256(installationSecret, currentHash))}. */
    public String computeSignature(String installationSecretBase64, String currentHash) {
        try {
            byte[] secretBytes = Base64.getDecoder().decode(installationSecretBase64);
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(secretBytes, "HmacSHA256"));
            byte[] signatureBytes = mac.doFinal(currentHash.getBytes(StandardCharsets.UTF_8));
            return Base64.getEncoder().encodeToString(signatureBytes);
        } catch (Exception e) {
            throw new IllegalStateException("HMAC-SHA256 signing failed", e);
        }
    }
}
